migrations/1606305812748-AddMakerBalanceCacheIndex.ts

```diff
@@ -0,0 +1,13 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddMakerBalanceCacheIndex1606305812748 implements MigrationInterface {
```

Comment:
Are you sure this index is necessary? Those are the primary key fields, so it shouldn't scan the whole table if you specify both.

Reply:
That's a great thought, let me remove the index and try running thr query

Reply:
https://gist.github.com/PirosB3/1548153af1e5b3939498cf42e44aef62

Sir, you are totally right! the `EXPLAIN `cost is the same.

```sql
api=# EXPLAIN SELECT * FROM maker_balance_chain_cache WHERE token_address = '0xfoo' AND maker_address IN ('a', 'b', 'c');
                                          QUERY PLAN
-----------------------------------------------------------------------------------------------
 Bitmap Heap Scan on maker_balance_chain_cache  (cost=4.17..11.29 rows=1 width=112)
   Recheck Cond: ((token_address)::text = '0xfoo'::text)
   Filter: ((maker_address)::text = ANY ('{a,b,c}'::text[]))
   ->  Bitmap Index Scan on "PK_13d68ce2e19b661c60a8499808c"  (cost=0.00..4.17 rows=3 width=0)
         Index Cond: ((token_address)::text = '0xfoo'::text)
(5 rows)
```

Will drop the new index

Reply:
@alexkroeger maybe this is irrelevant and it's over-optimizing (not a PG expert), but I'm surprised that the query that I run only partially hits the index.

```
api=# EXPLAIN SELECT * FROM maker_balance_chain_cache WHERE token_address = '0xfoo' AND maker_address IN ('a', 'b', 'c') AND maker_address IS NOT NULL;
                                                   QUERY PLAN
----------------------------------------------------------------------------------------------------------------
 Index Scan using maker_balance_chain_cache_idx on maker_balance_chain_cache  (cost=0.28..8.30 rows=1 width=56)
   Index Cond: (((token_address)::text = '0xfoo'::text) AND (maker_address IS NOT NULL))
   Filter: ((maker_address)::text = ANY ('{a,b,c}'::text[]))
(3 rows)
```

Is there any reason why PG does a linear scan (Which is what I interpret as a `Filter`) after indexing on `token_address`?

Reply:
I think the `maker_address is not null` may be throwing it off. Primary keys can't be null so you can get rid of that.

Reply:
oh sorry, I added that to see if I could fool the planner to use the index. The query plan is (nearly) the same when running: `SELECT * FROM maker_balance_chain_cache WHERE token_address = '0xfoo' AND maker_address IN ('a', 'b', 'c')`